man/en/jubaconfig.8
.\" Man page generated from reStructuredText.
.
.TH "JUBACONFIG" "8" " " "" "Jubatus"
.SH NAME
jubaconfig \- jubaconfig
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
jubaconfig \-\-cmd <command> [options ...]
.ft P
.fi
.UNINDENT
.UNINDENT
.SH DESCRIPTION
.sp
In distributed environment, \fBjubaconfig\fP manages the configuration files of Jubatus servers that are registered on ZooKeeper.
.SS Options
.INDENT 0.0
.IP \(bu 2
\fB[]\fP indicates the default value.
.IP \(bu 2
\fB<zookeeper_list>\fP is a comma\-separated list of \fBhost:port\fP of ZooKeeper servers (e.g., \fB10.0.0.1:2181,10.0.0.2:2181,10.0.0.3:2181\fP).
No spaces between values are allowed.
If you have only one ZooKeeper server, just specify \fBhost:port\fP without comma (e.g., \fB10.0.0.1:2181\fP).
.UNINDENT
.INDENT 0.0
.TP
.B \-c <command>, \-\-cmd <command>
Specify the action to perform on configuration files.
\fB<command>\fP should be one of the following.
.TS
center;
|l|l|.
_
T{
Command
T}	T{
Description
T}
_
T{
write
T}	T{
Register configuration file on the local file system to ZooKeeper
T}
_
T{
read
T}	T{
Display configuration file registered on ZooKeeper
T}
_
T{
delete
T}	T{
Remove configuration file registered on ZooKeeper
T}
_
T{
list
T}	T{
List configuration file registered on ZooKeeper
T}
_
.TE
.UNINDENT
.INDENT 0.0
.TP
.B \-f <file>, \-\-file <file>
Path of the configuration file to register to ZooKeeper.
.sp
Effective only when used with \fB\-\-cmd write\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-t <type>, \-\-type <type>
Type of the server program (e.g., \fBclassifier\fP, \fBrecommender\fP, ...).
.sp
Effective only when used with one of \fB\-\-cmd write\fP, \fB\-\-cmd read\fP or \fB\-\-cmd delete\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-n <name>, \-\-name <name>
The instance name, which is a value to uniquely identify a task in the ZooKeeper cluster.
.sp
Effective only when used with one of \fB\-\-cmd write\fP, \fB\-\-cmd read\fP or \fB\-\-cmd delete\fP\&.
.UNINDENT
.INDENT 0.0
.TP
.B \-z <zookeeper_list>, \-\-zookeeper <zookeeper_list>
List of ZooKeeper server(s).
.sp
If not specified, environment variable \fBZK\fP will be used.
.UNINDENT
.INDENT 0.0
.TP
.B \-d, \-\-debug
This option is deprecated and is no longer be used.
.UNINDENT
.INDENT 0.0
.TP
.B \-?, \-\-help
Print the brief usage of the command.
.UNINDENT
.SH AUTHOR
PFN & NTT
.SH COPYRIGHT
2011-2017 PFN & NTT
.\" Generated by docutils manpage writer.
.


ChangeLog.rst
Release 1.0.2 - 2017/02/27
--------------------------

* Changes to Supported Systems
    * Ubuntu 16.04 LTS is now officially supported (#1180, #1181)

Release 1.0.1 - 2016/12/26
--------------------------

* Improvements
    * Support OpenCV 2.0 in image extraction plugin (#1168, #1174, #1175)
    * fv_converter plugin loading code moved to Jubatus Core (#1171)
    * Improve Dockerfile (#1170, #1172, #1173)

Release 1.0.0 - 2016/10/31
--------------------------

* New Features
    * Add image extraction plug-in (#843, #1150)
    * Add configuration validation feature (#973, #1144)

* Improvements
    * clustering: change APIs to accept ID-based datum (#1155)
    * Update Dockerfile to use Ubuntu 14.04 (#1146)
    * Improve ``--daemon`` option of server and proxy to detach from console (#1072, #1148)
    * classifier: models are now compatible in standalone and distributed mode (#469, #1153)
    * Change the default log level to INFO instead of DEBUG (#1152)
    * Update example configurations (#1165)
    * Update manpages (#1167)
    * Update packaging tools (#1157)

* Bug fixes
    * graph: fix IDs being reused when loading model file to other server instance in standalone mode (#1149, #1156)

Release 0.9.4 - 2016/08/29
--------------------------

* Improvements
    * Improve error detection of log4cxx config file (#891, #1142)
    * Print DEBUG logs by default (#947, #1143)
    * Support printing logs in UTF-8 (#1078, #1140)
    * Improve comment in wscript (#1138, #1139)

* Bug fixes
    * Fix missing locks in mixer (#1092, #1135)
    * Fix RPM packaging tool to work on RHEL 6.x (#1137)

Release 0.9.3 - 2016/07/25
--------------------------

* Improvements
    * Improve behavior when ZooKeeper log files cannot be reopened (#1026, #1133)
    * Remove Glibc C++ debug flag in debug mode (#910, #1136)
    * Improve packaging tools to remove dependency to Google Code (#980, #1130)

* Bug fixes
    * weight: fix to display config on startup (#1131, #1132)
    * Fix missing lock in get_status of linear_mixer (#1093, #1134)

Release 0.9.2 - 2016/06/27
--------------------------

* Improvements
    * Update README.rst (#1126, #1129)

* Bug fixes
    * Fix `weight_client.hpp` not included in `client.hpp` (#1127, #1128)

Release 0.9.1 - 2016/05/30
--------------------------

* New feature
    * weight: server that returns raw results of fv_converter (#1105, #1116)

* Improvements
    * Add `default.json` config files (#1117, #1120)
    * Add example config file for Anomaly (lof with unlearner) (#1118)
    * Update manpage (#1123)

* Bug fixes
    * Fix thread safetiness issue in save/load RPC (#1113, #1115)
    * Fix `dl` link error on some distributions (#1121, #1122)

Release 0.9.0 - 2016/04/25
--------------------------

* New feature
    * `jubakit <https://github.com/jubatus/jubakit>`_: Python module to easily evaluate precision of models (#1044)

* Changes to Supported Systems
    * Ubuntu 14.04 LTS is now officially supported (#1071)

* Improvements
    * classifier: improve get_labels API to return number of trained instances (#1109)
    * Add example configuration file for Nearest Neighbor (multi-threaded LSH) (#1111)
    * Add example configuration file for Classifier (cosine / euclidean) (#1108)
    * Add example configuration file for Recommender (euclid_lsh with unlearner) (#1108)
    * Configuration in model is used when loading model file on startup (#1081, #1107)
    * Improve compatibility with FreeBSD 9 (#1099)
    * Update manpages (#1112)
    * Code cleanup (#1102)

* Bug fixes
    * Fix jenerator test (#1103)

Release 0.8.9 - 2016/03/28
--------------------------

* Improvements
    * Support extracting specific part-of-speech in MeCab plugin (#1073, #1098)
    * Improved proxy to log timeout errors (#1048, #1095)

* Bug fixes
    * Fix "jubactl -c save" not working (#1094)

Release 0.8.8 - 2016/02/29
--------------------------

* Improvements
    * Support extracting base form in MeCab plugin (#1074, #1085)
    * Compare configuration string semantically on load (#1080, 1087)
    * Add example configuration file for Recommender (inverted_index_euclid with unlearner) (#1090)
    * Fix email address of DEB package (#1082, #1088)

* Bug fixes
    * Fix printing logs multiple times after receiving SIGHUP (#962, #1084, #1086)

Release 0.8.7 - 2016/01/25
--------------------------

* Improvements
    * Add example configuration file for Recommender (MinHash with unlearner) (#1077)
    * Add example configuration file for Recommender and Anomaly (inverted_index_euclid) (#1079)

Release 0.8.6 - 2015/12/25
--------------------------

* Improvements
    * Support morpheme n-gram in mecab plugin (#1054, #1070)

* Bug fixes
    * Fix typos in exception and log messages (#1068, #1069)

Release 0.8.5 - 2015/11/30
--------------------------

* Improvements
    * improve client tests to work on Python 3 (#1043, #1063)
    * improve build script (#1064)
    * recommender: fix typo of calc_similarity function name (#1065)
    * recommender: add example config for LSH with unlearner (#1066)
    * update man pages (#1067)

* Bug fixes
    * Fix RPM package errors (#1059, #1060, #1061, #1062)

Release 0.8.4 - 2015/10/26
--------------------------

* Improvements
    * nearest_neighbor: improve analysis methods performance by eliminating giant lock (#1051)

* Bug fixes
    * Fix loading model files via command line option not working (#1055, #1057)

Release 0.8.3 - 2015/09/28
--------------------------

* Improvements
    * clustering: add seed option to example configuration files (#1047)
    * Fix packaging tools to use new oniguruma URL (#1045, #1046)

* Bug fixes
    * nearest_neighbor: fix argument type (#1042, #1049, #1050)

Release 0.8.2 - 2015/08/31
--------------------------

* Improvements
    * Add Dockerfile (#904)
    * Support add aggregator in jenerator (#1022, #1040)
    * Migrate ux-trie packaging from Google Code to GitHub (#1037)

* Bug fixes

jubatus/server/cmd/jubaconfig.cpp
// Jubatus: Online machine learning framework for distributed environment
// Copyright (C) 2012 Preferred Networks and Nippon Telegraph and Telephone Corporation.
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License version 2.1 as published by the Free Software Foundation.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA

#include <locale.h>

#include <cstring>
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <utility>
#include <vector>

#include "jubatus/util/lang/function.h"

#include "jubatus/core/common/exception.hpp"
#include "../third_party/cmdline/cmdline.h"
#include "../common/config.hpp"
#include "../common/membership.hpp"
#include "../common/zk.hpp"


static const std::string PROGNAME(JUBATUS_APPNAME "config");
static const std::string VERSION(JUBATUS_VERSION);

using std::string;
using std::cout;
using std::endl;
using std::vector;

void usage();
void set_config(
    const string& zkhosts,
    const string& type,
    const string& name,
    const string& configfile);
void delete_config(
    const string& zkhosts,
    const string& type,
    const string& name);
void get_config(const string& zkhosts, const string& type, const string& name);
void get_configs(const string& zkhosts);
void print_config(
    jubatus::server::common::lock_service& z,
    const string& type,
    const string& name);
void get_all_config_paths(
    jubatus::server::common::lock_service& z,
    std::vector<std::pair<string, string> >& ret);


int main(int argc, char** argv)
try {
  // Explicitly set the current locale to support Unicode output in logs.
  ::setlocale(LC_ALL, "");

  // Configures the logger.
  // We don't provide logging configuration feature for command line tools;
  // just print logs to standard output.
  jubatus::server::common::logger::configure();
  if (!jubatus::server::common::logger::is_configured()) {
    std::cerr << "failed to configure logger" << std::endl;
    ::exit(1);
  }

  cmdline::parser p;
  p.add<std::string>("cmd", 'c',
      "command to operate config (write|read|delete|list)", true);
  p.add<std::string>("file", 'f', "[write] config file to set", false, "");
  p.add<std::string>("type", 't',
      "[write|read|delete] learning machine type", false, "");
  p.add<std::string>("name", 'n',
      "[write|read|delete] learning machine name", false, "");
  p.add<std::string>("zookeeper", 'z',
      "ZooKeeper location environment: 'ZK' is available instead", false);

  p.add("debug", 'd', "debug mode (obsolete)");

  p.parse_check(argc, argv);

  string zk;
  char* zkenv = std::getenv("ZK");

  if (p.get<std::string>("zookeeper") != "") {
    zk = p.get<std::string>("zookeeper");
  } else if (zkenv) {
    zk = zkenv;
  } else {
    cout << "can't get ZK location: "
        << "set 'ZK' environment or specify '-z <somezkaddrs>" << endl;
    cout << p.get<std::string>("zookeeper") << endl;
    cout << p.usage() << endl;
    exit(1);
  }

  string cmd = p.get<std::string>("cmd");
  string type = p.get<std::string>("type");
  string name = p.get<std::string>("name");

  if (cmd == "write" || cmd == "delete" || cmd == "read") {
    if (type == "" || name == "") {
      cout << "can't excute " << cmd
          << " without type and name specified" << endl;
      cout << "type: " << type << " name: " << name << endl;
      cout << p.usage() << endl;
      exit(1);
    }

    if (cmd == "write") {
      set_config(zk, type, name, p.get<std::string>("file"));
    }

    if (cmd == "delete") {
      delete_config(zk, type, name);
    }

    if (cmd == "read") {
      get_config(zk, type, name);
    }
  } else if (cmd == "list") {
    get_configs(zk);
  } else {
    cout << "unknown cmd: " << cmd << endl;
    cout << p.usage() << endl;
    exit(1);
  }

  return 0;
} catch (const jubatus::core::common::exception::jubatus_exception& e) {
  cout << e.diagnostic_information(true) << endl;
  exit(1);
}

void set_config(
    const string& zkhosts,
    const string& type,
    const string& name,
    const string& configfile) {
  jubatus::util::lang::shared_ptr<jubatus::server::common::lock_service> ls_(
      jubatus::server::common::create_lock_service(
          "zk", zkhosts, 10, "/dev/null"));

  jubatus::server::common::prepare_jubatus(*ls_, type, name);

  string config;
  jubatus::server::common::config_fromlocal(configfile, config);
  jubatus::server::common::config_tozk(*ls_, type, name, config, configfile);
}

void delete_config(
    const string& zkhosts,
    const string& type,
    const string& name) {
  jubatus::util::lang::shared_ptr<jubatus::server::common::lock_service> ls_(
      jubatus::server::common::create_lock_service(
          "zk", zkhosts, 10, "/dev/null"));

  jubatus::server::common::remove_config_fromzk(*ls_, type, name);
}

void get_config(const string& zkhosts, const string& type, const string& name) {
  jubatus::util::lang::shared_ptr<jubatus::server::common::lock_service> ls_(
      jubatus::server::common::create_lock_service(
          "zk", zkhosts, 10, "/dev/null"));

  print_config(*ls_, type, name);
}

void get_configs(const string& zkhosts) {
  jubatus::util::lang::shared_ptr<jubatus::server::common::lock_service> ls_(
      jubatus::server::common::create_lock_service(
          "zk", zkhosts, 10, "/dev/null"));

  std::vector<std::pair<string, string> > paths;
  get_all_config_paths(*ls_, paths);

  for (size_t i = 0; i < paths.size(); i++) {
    cout << "type: " << paths[i].first << ", name: " << paths[i].second << endl;
    print_config(*ls_, paths[i].first, paths[i].second);
    cout << '\n';
  }
}

void print_config(
    jubatus::server::common::lock_service& z,
    const string& type,
    const string& name) {
  string config;
  jubatus::server::common::config_fromzk(z, type, name, config);
  cout << config;
}

void get_all_config_paths(
    jubatus::server::common::lock_service& z,
    std::vector<std::pair<string, string> >& ret) {
  ret.clear();

  std::vector<std::string> types;
  if (!z.list(jubatus::server::common::CONFIG_BASE_PATH, types)
      || types.empty()) {
    return;
  }

  for (size_t i = 0; i < types.size(); i++) {
    std::vector<std::string> names;
    std::string path;
    path = jubatus::server::common::CONFIG_BASE_PATH + '/' + types[i];
    if (!z.list(path, names) || names.empty()) {
      continue;
    }

    for (size_t j = 0; j < names.size(); j++) {
      std::string config_path;
      config_path = path + '/' + names[j];
      ret.push_back(make_pair(types[i], names[j]));
    }
  }
}
